optuna/_multi_objective.py

```diff
@@ -8,7 +8,46 @@
 from optuna.trial import TrialState
 
 
-def _get_pareto_front_trials(study: "optuna.study.BaseStudy") -> List[FrozenTrial]:
+def _get_pareto_front_trials_2d(study: "optuna.study.BaseStudy") -> List[FrozenTrial]:
```

Comment:
How about sorting the trials themself, and using `_dominates` function as done in `_get_pareto_front_trials_nd`?
```python
    pareto_front = []
    trials = [trial for trial in study.trials if trial.state == TrialState.COMPLETE]
    trials.sort(key=lambda t: _normalize_value(t.values[1], study.directions[1]))
    trials.sort(key=lambda t: _normalize_value(t.values[0], study.directions[0]))

    last_nondominated_trial: Optional[FrozenTrial] = None
    for i in range(len(trials)):
        if i == 0:
            pareto_front.append(trials[0])
            last_nondominated_trial = trials[0]
            continue

        assert last_nondominated_trial is not None
        if not _dominates(last_nondominated_trial, trials[i], study.directions):
            pareto_front.append(trials[i])
            last_nondominated_trial = trials[i]
```
This code does not preserve the order of all trials and pareto front trials. so it will be fixed.

Reply:
I like the idea of using `_dominates` to simplify the code. I did something similar to what you suggested but because, as you point out, the order of trials needs to be preserved when filtering, I am still using `mask`.

Reply:
To preserve the order of trials, we only need to sort the Pareto front trials with its `trial.number`. Could you remove the `mask`?

Reply:
Nice catch. Done.